### ISSUE TITLE
Add support for embassy UCPD

### ISSUE THREAD

**Issue by elagil** [fmckeogh/usb-pd-rs#22]
Hello,

thank you for writing this library!

I am in the process of using this with the UCPD peripheral on STM32 (I will test with STM32G4) on embassy for implementing a `Sink`.

I have some ideas in mind for changes to the sink's `Driver` trait, because I have some trouble to use UCPD with it as is:
- Offer an alternative for `poll`/`get_pending_message`, because in UCPD, both steps are handled with a call to `receive().await` on the PD phy instance. For example, add something like `receive_message`. Maybe this could also be refactored on the FUSB302?
- Add `Result` return type to `send_message` and `receive_message` and propagate errors to the user application (might drop the driver and make a new one on error)

Some general ideas:
- Remove `did_change_protocol` from the `Driver` trait and instead check if the `state` of the `Driver` changed from within the `Sink`.

What do you think?

**Comment by elagil** [comment]
I added a PR for showing those ideas.

**Comment by fmckeogh** [comment]
Thanks so much, yeah that all sounds great:) I can't test any FUSB301B regressions, but I'll merge the PR and fix anything that breaks later, if that helps unblock your work.

**Comment by elagil** [comment]
It's not blocking me at all, I can just use my fork :D

My intention is to make your library a bit more generic, if you agree to the proposed changes. Unfortunately, I also have no way of testing on FUSB301B.

I will likely open another PR later with some more things that we can discuss.

**Comment by fmckeogh** [comment]
@elagil I’m fully on board, I’d love to make this correctly-abstracted so that core PD logic can be shared with as many device implementations as possible.

> I will likely open another PR later with some more things that we can discuss.

Sounds great!

**Comment by fmckeogh** [comment]
@elagil IIRC, those are required for timing out if negotiations don't occur within 500ms, and toggling CC lines every 20ms

**Comment by elagil** [comment]
@fmckeogh I am asking because they are not actually used in the library. Am I missing something?

**Comment by fmckeogh** [comment]
@elagil My apologies, I thought you were referring to the `Instant` timer logic, if you mean the `usb_pd::timers` module, @bentwire added that I assumed for future USB PD standard work

**Comment by bentwire** [comment]
Yes, sorry I've been out of the loop.  Only doing this as a hobby and life got in the way.  

Those timers were supposed to be for the pd state machine yes.

**Comment by okhsunrog** [comment]
@elagil do you have a template UCPD project for stm32g4 that you can share? I didn't find any on your github page. I'd really appreciate that

**Comment by elagil** [comment]
@okhsunrog Yes, but it only works with another new branch of this project. I will push it when it's ready, but it needs a bit more work.

**Comment by elagil** [comment]
@okhsunrog I ended up releasing my own crate, because I would have had to modify this one too much and break compatibility with FUSB302B. See https://crates.io/crates/usbpd. An example for STM32G4 is here: https://github.com/elagil/usbpd/tree/main/examples/embassy-stm32-g431cb

**Comment by okhsunrog** [comment]
@elagil Thanks! I have a B-G474E-DPOW1 dev board, I'll test on it. I was actually thinking ideally about a more generic crate, that could support FUSB302B, STM32's UCPD and CH32X035 chips, that would be great. How hard it would be?

**Comment by elagil** [comment]
> I was actually thinking ideally about a more generic crate

The crate is generic over the driver implementation. Let's discuss [here](https://github.com/elagil/usbpd/issues/7)

**Comment by fmckeogh** [comment]
@elagil Do you want the `usb-pd` crates.io name? I might archive this repo and send people to yours; I fixed the problem I was having in hardware so no longer have much need for this crate and probably won't develop it further.

**Comment by elagil** [comment]
@fmckeogh Thanks for the offer. If you really have no use for it, that would be really nice! `usbpd` looks a bit cryptic as a crate name.

**Comment by fmckeogh** [comment]
@elagil Invite sent:)
